File: NeetCode/2-D_Dynamic_Programming/Longest_Common_Subsequence/longest__common__subsequence.py
from typing import Optional, List


class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        seen = {}
        seen[("", "")] = 0

        set1 = set(text1)
        set2 = set(text2)
        check = len(set1.intersection(set2))
        if check in (0, 1):
            return check

        def rec(t1, t2):
            if (t1, t2) in seen:
                return seen[(t1, t2)]
            ptr1, ptr2 = 0, 0
            if ptr1 == len(t1) or ptr2 == len(t2):
                return 0
            if t1[ptr1] == t2[ptr2]:
                tmp = 1 + rec(t1[ptr1 + 1 :], t2[ptr2 + 1 :])
                seen[(t1, t2)] = tmp
                return tmp
            first = rec(t1[ptr1 + 1 :], t2)
            second = rec(t1, t2[ptr2 + 1 :])
            tmp = max(first, second)
            seen[(t1, t2)] = tmp
            return tmp

        return rec(text1, text2)


if __name__ == "__main__":
    sol = Solution()
    text1 = "cat"
    text2 = "dog"
    print(sol.longestCommonSubsequence(text1, text2))
    print("Running Solution...")

File: NeetCode/2-D_Dynamic_Programming/Longest_Common_Subsequence/Longest_Common_Subsequence.cpp
// Longest_Common_Subsequence.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
private:
    struct PairHash
    {
        size_t operator()(const pair<string, string> &p) const
        {
            size_t h1 = hash<string>{}(p.first);
            size_t h2 = hash<string>{}(p.second);
            return h1 ^ (h2 << 1);
        }
    };
    unordered_map<pair<string, string>, int, PairHash> seen;

public:
    int longestCommonSubsequence(string text1, string text2)
    {
        seen[{"", ""}] = 0;

        return rec(text1, text2, 0, 0);
    }
    int rec(string &a, string &b, int ptr1, int ptr2)
    {

        if (seen.count({a.substr(ptr1), b.substr(ptr2)}))
        {
            return seen[{a.substr(ptr1), b.substr(ptr2)}];
        }
        if (a.substr(ptr1).size() == 0 || b.substr(ptr2).size() == 0)
        {
            return 0;
        }

        if (a[ptr1] == b[ptr2])
        {
            seen[{a.substr(ptr1), b.substr(ptr2)}] = 1 + rec(a, b, ptr1 + 1, ptr2 + 1);
            return seen[{a.substr(ptr1), b.substr(ptr2)}];
        }
        int first = rec(a, b, ptr1 + 1, ptr2);
        int second = rec(a, b, ptr1, ptr2 + 1);
        seen[{a.substr(ptr1), b.substr(ptr2)}] = max(first, second);
        return seen[{a.substr(ptr1), b.substr(ptr2)}];
    }
};
